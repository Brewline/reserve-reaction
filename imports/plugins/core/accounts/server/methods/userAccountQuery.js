import { Meteor } from "meteor/meteor";

/**
 * @name userAccountQuery
 * @method
 * @memberof Accounts/NoMeteorQueries
 * @summary query the Accounts collection and return user account data
 * @param {Object} context - an object containing the per-request state
 * @param {String} id - id of user to query
 * @return {Object} user account object
 */
export async function userAccountQuery(context, id) {
  const { collections, userHasPermission, userId } = context;
  const { Accounts } = collections;

  const userAccount = await Accounts.findOne({ _id: id });
  if (!userAccount) throw new Meteor.Error("not-found", "No account found");

  // Check to make sure current user has permissions to view queried user
  if (userId !== id && !userHasPermission(["reaction-accounts"], userAccount.shopId)) {
    throw new Meteor.Error("access-denied", "User does not have permission");
  }

  return userAccount;
}
