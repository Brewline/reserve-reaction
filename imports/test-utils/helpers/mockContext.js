const mockContext = {
  collections: {},
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "Logs",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Sms",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
