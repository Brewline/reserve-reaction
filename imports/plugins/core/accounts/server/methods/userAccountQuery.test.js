import mockContext from "/imports/test-utils/helpers/mockContext";
import { userAccountQuery } from "./userAccountQuery";

const fakeAccountId = "FAKE_USER_ID_FOR_QUERY";
const fakeAccount = { _id: fakeAccountId, shopId: "ACCOUNT_SHOP_ID" };

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws if account not found", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
  await expect(userAccountQuery(mockContext, fakeAccountId)).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.collections.Accounts.findOne).toHaveBeenCalledWith({ _id: fakeAccountId });
});

test("throws if userHasPermission returns false and the user ID is not the context user ID", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  await expect(userAccountQuery(mockContext, fakeAccountId)).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["reaction-accounts"], fakeAccount.shopId);
});

test("returns the account without calling userHasPermission if the user ID is the context user ID", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(false);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  const result = await userAccountQuery(mockContext, mockContext.userId);
  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
  expect(result).toEqual(fakeAccount);
});

test("returns the account if the user ID is not the context user ID but userHasPermission returns true", async () => {
  mockContext.userHasPermission.mockReturnValueOnce(true);
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(fakeAccount);
  const result = await userAccountQuery(mockContext, fakeAccountId);
  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["reaction-accounts"], fakeAccount.shopId);
  expect(result).toEqual(fakeAccount);
});
