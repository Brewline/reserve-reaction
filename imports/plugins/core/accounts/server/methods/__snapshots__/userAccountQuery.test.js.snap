// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if account not found 1`] = `"not-found"`;

exports[`throws if userHasPermission returns false and the user ID is not the context user ID 1`] = `"access-denied"`;
